### ISSUE TITLE
Inconsistent type inference

### ISSUE THREAD

**Issue by Krzmbrzl** [microsoft/pyright#4671]
**Describe the bug**
If you run pyright on the code snippet provided below, it will error on the argument for the first `np.array` call, but is perfectly content with the second. According to the `coc-pyright` plugin, this is because in the first case the argument is deduced to be of type `list[list[int | float]]`. However, the type of `arg` is instead deduced to be `list[list[Unknown]]` and (I believe) therefore no additional checking is done for the second `np.array` invocation (thus, no error produced).

The snippet doesn't produce any runtime errors.

**To Reproduce**
Run pyright on 
```python
import numpy as np

array1 = np.array([[1,2,3.0]])
arg = [[1,2,3.0]]
array2 = np.array(arg)
```
with Numpy 1.24.2 (maybe slightly older also works - the issue did not occur with older Numpy versions. Potentially, this is related to Numpy's recent addition of type stubs)

Running `pyright test.py` with pyright v1.1.295 (installed via pip) yields
```
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.295
/tmp/test.py
  /tmp/test.py:3:20 - error: Argument of type "list[list[int | float]]" cannot be assigned to parameter "object" of type "_ArrayLike[_SCT@array]" in function "array"
    "list[int | float]" is incompatible with protocol "_SupportsArray[dtype[Unknown]]"
      "__array__" is not present (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.907sec
```

**Expected behavior**
Ideally, no error should be raised. Though, this might also be related to how Numpy has implemented their type constraints. More importantly, I would expect the type of the list `[[1,2,3.0]]` to be deduced to the same type, regardless of the context in which it appears.

**VS Code extension or command-line**
CLI, version 1.1.295

**Comment by Krzmbrzl** [comment]
Note: The reported error in the first `np.array` invocation can be fixed by explicitly specifying the `dtype` for the array like so:
```python
array1 = np.array([[1,2,3.0]], dtype=float]])
```

Also note that instantiating a vector instead of a matrix does not report any errors, even without `dtype`:
```python
array3 = np.array([1,2,3.0])
```

**Comment by erictraut** [comment]
This is working as designed. If you configure pyright to use strict list inference, the error will go away. Set `strictListInference` to true in your configuration.

**Comment by Krzmbrzl** [comment]
Could you elaborate a little bit on why this is intended? To me it seems like the exact same list gets inferred differently, albeit in seemingly rather similar contexts :thinking:

**Comment by erictraut** [comment]
In one case, [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types) is used (i.e. the inference result is impacted by the expected type as dictated by the corresponding parameter). In the other it's not.

**Comment by Krzmbrzl** [comment]
Ah, neat. Didn't know something like this existed. Thanks!
